### DISCUSSION TITLE
Getting Jpg image from SD_card to sprite (or memory) - ESP32

### DISCUSSION THREAD

**Discussion by SL06** [Bodmer/TFT_eSPI#2199] [Q&A - Sprites] [answered]
Hi, I am new to TFT.  I have a set of  9  openstreetmap tiles 256 x 256 pixels  Jpeg, stored ion a SD_card and would like to put them in memory to be able to move the displayed portion of the map on the TFT screen smoothly. I have tried to put the transfert the images in a series sprites, but was unsuccessful. I am not shure if its the proper approach.  Here is the code for simple jpeg file, based on the Jpg_decoder SD__card esemple.  

`// Example for library:
// https://github.com/Bodmer/TJpg_Decoder

// This example if for an ESP8266 or ESP32, it renders a Jpeg file
// that is stored in a SD card file. The test image is in the sketch
// "data" folder (press Ctrl+K to see it). You must save the image
// to the SD card using you PC.

// Include the jpeg decoder library
#include <TJpg_Decoder.h>

// Include SD
#define FS_NO_GLOBALS
#include <FS.h>
#ifdef ESP32
  #include "SPIFFS.h" // ESP32 only
#endif

//#define SD_CS   4

// Include the TFT library https://github.com/Bodmer/TFT_eSPI
#include "SPI.h"
#include <TFT_eSPI.h>              // Hardware-specific library
TFT_eSPI tft = TFT_eSPI();         // Invoke custom library

TFT_eSprite spr = TFT_eSprite(&tft);

// This next function will be called during decoding of the jpeg file to
// render each block to the TFT.  If you use a different TFT library
// you will need to adapt this function to suit.
bool tft_output(int16_t x, int16_t y, uint16_t w, uint16_t h, uint16_t* bitmap)
{
   // Stop further decoding as image is running off bottom of screen
  if ( y >= 240 ) return 0; // sprite height - block size

  //Serial.println(w);  // 16... last 12

  // This function will clip the image block rendering automatically at the TFT boundaries
  //tft.pushImage(x, y, w, h, bitmap);
  if  ( x <= 239) {   // sprite height - block size
  spr.pushImage(x, y, w, h, bitmap);}
  

  // This might work instead if you adapt the sketch to use the Adafruit_GFX library
  // tft.drawRGBBitmap(x, y, bitmap, w, h);

  // Return 1 to decode next block
  return 1;
}

void setup()
{
  Serial.begin(115200);
  Serial.println("\n\n Testing TJpg_Decoder library");

  // Initialise SD before TFT
  //if (!SD.begin(SD_CS)) {
  if (!SD.begin()) {
    Serial.println(F("SD.begin failed!"));
    while (1) delay(0);
  }
  Serial.println("\r\nInitialisation done.");

  // Initialise the TFT
  tft.begin();
  tft.setTextColor(0xFFFF, 0x0000);
  tft.fillScreen(TFT_BLACK);
  tft.setSwapBytes(true); // We need to swap the colour bytes (endianess)

  // The jpeg image can be scaled by a factor of 1, 2, 4, or 8
  TJpgDec.setJpgScale(1);

  // The decoder must be given the exact name of the rendering function above
  TJpgDec.setCallback(tft_output);

   spr.createSprite(256, 256);
}

void loop()
{
  tft.fillScreen(TFT_RED);

  // Time recorded for test purposes
  uint32_t t = millis();

  // Get the width and height in pixels of the jpeg if you wish
  uint16_t w = 0, h = 0;
  TJpgDec.getSdJpgSize(&w, &h, "/EagleEye.jpg");
  Serial.print("Width = "); Serial.print(w); Serial.print(", height = "); Serial.println(h);

  // Draw the image, top left at 0,0
  //TJpgDec.drawSdJpg(0, 0, "/panda.jpg");

  TJpgDec.drawSdJpg(0, 0, "/EagleEye.jpg");
  // How much time did rendering take
  spr.pushSprite(50,50);

  t = millis() - t;
  Serial.print(t); Serial.println(" ms");

  // Wait before drawing again
  delay(10000);
}
`

**Comment by Bodmer** [comment, marked as answer]
I suspect that sprite creation fails since 128kbytes of RAM will be required. If you have onboard PSRAM then it must be enabled. Try a smaller sprite, say 256x64.

**Comment by Bodmer** [comment]
To check for successful creation you can use code like this:
```
  if (img.createSprite(100, 100) == nullptr) Serial.println("Sprite NOT created, insufficient RAM!");
  else Serial.println("YES! Sprite created.");
```

**Comment by SL06** [comment]
Thanks for those Email.



Setting the colour depth  to 8 bits  makes the sprite work.   :  spr.setColorDepth(8);   256 x 256 bytes is OK
With  16 bit, the size is limited to about half (not completely tested).   
  
Setting the PSRAM  to true did not help.   spr.setAttribute(PSRAM_ENABLE, true);


My board is an ESPDUINO-32 HW-729 with the ESP32-D0WDQ6 (revision 


I have checked the PRAM with the following command and it told  PSRAM size =  0, although 

Serial.printf("PSRAM Total heap %d, PSRAM Free Heap %d\n",ESP.getPsramSize(),ESP.getFreePsram());



This code also give me PSRAM not available….


 if(psramInit()){
        Serial.println("\nPSRAM is correctly initialized");
        }else{
        Serial.println("PSRAM not available");
        }

 I have the latest arduino version and ESP-IDE.  I assume that the PSRAM is activated by default and that I don’t have to upload any special library other may be #include <Arduino.h>…

I will keep looking.
Any suggestion is appreciated.



Sylvain

**Comment by SL06** [comment]
Just found this.  My ESP32 version does not have PSRAM according to this. https://www.upesy.com/blogs/tutorials/get-more-ram-on-esp32-with-psram <https://www.upesy.com/blogs/tutorials/get-more-ram-on-esp32-with-psram>

I would need Oe of those. 
ESP32-WROVER-B
ESP32-WROVER


Thanks again.
Sylvain

**Comment by Reyansh-Niranjan** [comment]
Hi,
Same issue, But can it be done without PSRAM?
My current one is soldered and found it out that it doesn't have PSRAM ON IT
soo..
any help?

**Comment by SL06** [comment]
I don’t know the answer but ChatGPT or Atlas GPT might.  I rely on them more and more for programming.

S

**Reply by Reyansh-Niranjan** [reply]
Ok Thanks!

**Comment by remysedlak** [comment]
I got images displaying from SD card on my Hosyond ESP32-3248S035 board. 8-bit sprites work fine (even up to 200x200), but 16-bit sprites fail with anything larger than like 160x160 - just runs out of RAM since this board has no PSRAM.
One weird issue though - after doing SD operations, my touchscreen stops responding. I have to restart the ESP32 to get touch working again. Not sure if it's specific to this board since the SD card, TFT, and resistive touch all share the same SPI bus.
